Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,207 @@
+def SimpleGraph.degMatrix [AddMonoidWithOne α] : Matrix V V α := Matrix.diagonal (G.degree ·)
+
+/-- `lapMatrix G R` is the matrix `L = D - A` where `D`is the degree
+  and `A` the adjacency matrix of `G`. -/
+def SimpleGraph.lapMatrix [AddGroupWithOne α] : Matrix V V α := G.degMatrix α - G.adjMatrix α
+
+variable {α}
+
+theorem isSymm_lapMatrix [AddGroupWithOne α] : (G.lapMatrix α).IsSymm := by
+  rw [Matrix.IsSymm, lapMatrix, degMatrix, transpose_sub, diagonal_transpose, transpose_adjMatrix]
+
+theorem lapMatrix_mulVec_const_eq_zero [Ring α] : mulVec (G.lapMatrix α) (fun _ ↦ 1) = 0 := by
+  ext1 i
+  simp_rw [lapMatrix, sub_mulVec, Pi.sub_apply, adjMatrix_mulVec_apply, degMatrix, mulVec,
+    dotProduct, diagonal]
+  simp
+
+theorem dotProduct_mulVec_adjMatrix [Ring α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.adjMatrix α) x = ∑ i : V, ∑ j : V, if G.Adj i j then x i * x j else 0 := by
+  unfold mulVec dotProduct
+  simp [mul_sum]
+
+theorem dotProduct_mulVec_degMatrix [CommRing α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.degMatrix α) x = ∑ i : V, G.degree i * x i * x i := by
+  unfold dotProduct degMatrix
+  simp only [mulVec_diagonal, ← mul_assoc, mul_comm (x _)]
+
+variable (α)
+
+theorem sum_adj_eq_degree [AddCommMonoidWithOne α] (i : V) :
+    (G.degree i : α) = ∑ j : V, if G.Adj i j then 1 else 0 := by
+  unfold degree neighborFinset neighborSet
+  rw [sum_boole, Set.toFinset_setOf]
+
+/-- Let $L$ be the graph Laplacian and let $x \in \mathbb{R}$, then
+ $$x^{\top} L x = \sum_{i \sim j} (x_{i}-x_{j})^{2}$$
+ where $\sim$ denotes the adjacency relation -/
+theorem lapMatrix_toLinearMap₂' [Field α] [CharZero α] (x : V → α) :
+    toLinearMap₂' (G.lapMatrix α) x x =
+    (∑ i : V, ∑ j : V, if G.Adj i j then (x i - x j)^2 else 0) / 2 := by
+  rw [toLinearMap₂'_apply']
+  unfold lapMatrix
+  rw [sub_mulVec]
+  simp only [dotProduct_sub]
+  rw [dotProduct_mulVec_degMatrix, dotProduct_mulVec_adjMatrix, ← sum_sub_distrib]
+  simp only [sum_adj_eq_degree, sum_mul, ← sum_sub_distrib, ite_mul, one_mul,
+             zero_mul, ite_sub_ite, sub_zero]
+  rw [← half_add_self (∑ x_1 : V, ∑ x_2 : V, _)]
+  conv_lhs => arg 1; arg 2; arg 2; intro i; arg 2; intro j; rw [if_congr (adj_comm G i j) rfl rfl]
+  conv_lhs => arg 1; arg 2; rw [Finset.sum_comm]
+  simp_rw [← sum_add_distrib, ite_add_ite]
+  congr 2 with i
+  congr 2 with j
+  ring_nf
+
+/-- The Laplacian matrix is positive semidefinite -/
+theorem isPosSemidef_lapMatrix [LinearOrderedField α] [StarOrderedRing α] [TrivialStar α] :
+    PosSemidef (G.lapMatrix α) := by
+  constructor
+  · rw [IsHermitian, conjTranspose_eq_transpose_of_trivial, isSymm_lapMatrix]
+  · intro x
+    rw [star_trivial, ← toLinearMap₂'_apply', lapMatrix_toLinearMap₂']
+    refine div_nonneg (sum_nonneg' fun i ↦ sum_nonneg' fun j ↦ ?_) two_pos.le
+    split
+    · apply sq_nonneg
+    · rfl
+
+theorem lapMatrix_toLinearMap₂'_apply'_zero_iff_adj [LinearOrderedField α] (x : V → α) :
```

Comment:
The name should have `eq` in there. Maybe `forall` too? `lapMatrix_toLinearMap₂'_apply'_eq_zero_iff_forall_adj`